samza-core/src/main/scala/org/apache/samza/job/local/ProcessJob.scala

```diff
@@ -19,113 +19,118 @@
-  var process: Process = null
-
-  def submit: StreamJob = {
-    jobStatus = Some(New)
-    val waitForThreadStart = new CountDownLatch(1)
+object ProcessJob {
+  private def createProcessBuilder(commandBuilder: CommandBuilder): ProcessBuilder = {
     val processBuilder = new ProcessBuilder(commandBuilder.buildCommand.split(" ").toList.asJava)
+    processBuilder.environment.putAll(commandBuilder.buildEnvironment)
+
+    // Pipe all output to this process's streams.
+    processBuilder.redirectOutput(ProcessBuilder.Redirect.INHERIT)
+    processBuilder.redirectError(ProcessBuilder.Redirect.INHERIT)
 
     processBuilder
-      .environment
-      .putAll(commandBuilder.buildEnvironment)
+  }
+}
+
+class ProcessJob(commandBuilder: CommandBuilder, val jobModelManager: JobModelManager) extends StreamJob with Logging {
+
+  import ProcessJob._
+
+  val lock = new Object
+  val processBuilder: ProcessBuilder = createProcessBuilder(commandBuilder)
+  var jobStatus: ApplicationStatus = New
+  var processThread: Option[Thread] = None
+
 
-    // create a non-daemon thread to make job runner block until the job finishes.
-    // without this, the proc dies when job runner ends.
-    val procThread = new Thread {
+  def submit: StreamJob = {
+    val threadStartCountDownLatch = new CountDownLatch(1)
+
+    // Create a non-daemon thread to make job runner block until the job finishes.
+    // Without this, the proc dies when job runner ends.
+    processThread = Some(new Thread {
       override def run {
-        process = processBuilder.start
-
-        // pipe all output to this process's streams
-        val outThread = new Thread(new Piper(process.getInputStream, System.out))
-        val errThread = new Thread(new Piper(process.getErrorStream, System.err))
-        outThread.setDaemon(true)
-        errThread.setDaemon(true)
-        outThread.start
-        errThread.start
-        waitForThreadStart.countDown
-        process.waitFor
-        jobCoordinator.stop
+        var processExitCode = -1
+        var process: Option[Process] = None
+
+        setStatus(Running)
+
+        try {
+          threadStartCountDownLatch.countDown
+          process = Some(processBuilder.start)
+          processExitCode = process.get.waitFor
+        } catch {
+          case _: InterruptedException => process foreach { p => p.destroyForcibly }
+          case e: Exception => error("Encountered an error during job start: %s".format(e.getMessage))
+        } finally {
+          jobModelManager.stop
+          setStatus(if (processExitCode == 0) SuccessfulFinish else UnsuccessfulFinish)
+        }
       }
-    }
+    })
 
-    procThread.start
-    waitForThreadStart.await
-    jobStatus = Some(Running)
+    processThread.get.start
+    threadStartCountDownLatch.await
     ProcessJob.this
   }
 
   def kill: StreamJob = {
-    process.destroyForcibly
-    jobStatus = Some(UnsuccessfulFinish)
+    if (getStatus == Running) {
```

Comment:
* It will be useful to log the action(kill/submit) at the start of the methods.
* I think it will be useful to log the state transitions triggered by each one of these methods. It will help debugging any bugs associated with these methods.

Reply:
Sure. Will add.